diff --git a/Makefile b/Makefile
index 83a00d9..6a2c4d8 100644
--- a/Makefile
+++ b/Makefile
@@ -19,6 +19,7 @@ define Package/luci-app-koolproxy
 	TITLE:=LuCI support for koolproxy
 	DEPENDS:=+openssl-util +ipset +dnsmasq-full +@BUSYBOX_CONFIG_DIFF +iptables-mod-nat-extra +wget
 	MAINTAINER:=paradislover
+	PKGARCH=all
 endef
 
 define Package/luci-app-koolproxy/description
@@ -74,21 +75,21 @@ define Package/luci-app-koolproxy/install
 	$(INSTALL_BIN) ./files/usr/share/koolproxy/koolproxyupdate $(1)/usr/share/koolproxy/koolproxyupdate
 	$(INSTALL_DATA) ./files/usr/share/koolproxy/adblock.conf $(1)/usr/share/koolproxy/adblock.conf
 	$(INSTALL_DATA) ./files/usr/share/koolproxy/dnsmasq.adblock $(1)/usr/share/koolproxy/dnsmasq.adblock
-ifeq ($(ARCH),mipsel)
-	$(INSTALL_BIN) ./files/bin/mipsel $(1)/usr/share/koolproxy/koolproxy
-endif
-ifeq ($(ARCH),mips)
-	$(INSTALL_BIN) ./files/bin/mips $(1)/usr/share/koolproxy/koolproxy
-endif
-ifeq ($(ARCH),i386)
-	$(INSTALL_BIN) ./files/bin/i386 $(1)/usr/share/koolproxy/koolproxy
-endif
-ifeq ($(ARCH),x86_64)
-	$(INSTALL_BIN) ./files/bin/x86_64 $(1)/usr/share/koolproxy/koolproxy
-endif
-ifeq ($(ARCH),arm)
-	$(INSTALL_BIN) ./files/bin/arm $(1)/usr/share/koolproxy/koolproxy
-endif
+#ifeq ($(ARCH),mipsel)
+#	$(INSTALL_BIN) ./files/bin/mipsel $(1)/usr/share/koolproxy/koolproxy
+#endif
+#ifeq ($(ARCH),mips)
+#	$(INSTALL_BIN) ./files/bin/mips $(1)/usr/share/koolproxy/koolproxy
+#endif
+#ifeq ($(ARCH),i386)
+#	$(INSTALL_BIN) ./files/bin/i386 $(1)/usr/share/koolproxy/koolproxy
+#endif
+#ifeq ($(ARCH),x86_64)
+#	$(INSTALL_BIN) ./files/bin/x86_64 $(1)/usr/share/koolproxy/koolproxy
+#endif
+#ifeq ($(ARCH),arm)
+#	$(INSTALL_BIN) ./files/bin/arm $(1)/usr/share/koolproxy/koolproxy
+#endif
 endef
 
 $(eval $(call BuildPackage,luci-app-koolproxy))
