diff --git a/Makefile.openssl b/Makefile.openssl
index bc35630..d54a636 100644
--- a/Makefile.openssl
+++ b/Makefile.openssl
@@ -2,7 +2,7 @@ CC?=gcc
 CXX?=g++
 CFLAGS+=-fexceptions -DOPENSSL=1 -O2
 CXXFLAGS=-fexceptions -DOPENSSL=1 -O2
-LIBS=sendmsg.o openssldl.o cJSON.o nonblocking.o sslbio.o ngrok.o main.o
+LIBS=global.o sendmsg.o openssldl.o cJSON.o nonblocking.o opensslbio.o ngrok.o main.o
 all:: ngrokc $(LIBS)
 ngrokc: $(LIBS)
 	$(CXX) -s $(LIBS) -o ngrokc -lssl -lcrypto -ldl
@@ -11,8 +11,8 @@ openssldl.o: openssldl.h
 global.o: global.h
 cJSON.o: cJSON.h
 nonblocking.o: nonblocking.h
-sslbio.o: sslbio.h
+opensslbio.o: opensslbio.h
 ngrok.o: ngrok.h
-main.o: sendmsg.h openssldl.h sslbio.h ngrok.h cJSON.h bytestool.h nonblocking.h global.h
+main.o: sendmsg.h openssldl.h opensslbio.h ngrok.h cJSON.h bytestool.h nonblocking.h global.h
 clean:
 	rm -f $(LIBS) ngrokc
