diff --git a/block.c b/block.c
index 75093f0..c9b8f64 100644
--- a/block.c
+++ b/block.c
@@ -697,31 +697,6 @@ static char* find_mount_point(char *block)
 	return point;
 }
 
-static int print_block_uci(struct probe_info *pr)
-{
-	if (!strcmp(pr->type, "swap")) {
-		printf("config 'swap'\n");
-	} else {
-		char *mp = find_mount_point(pr->dev);
-
-		printf("config 'mount'\n");
-		if (mp) {
-			printf("\toption\ttarget\t'%s'\n", mp);
-			free(mp);
-		} else {
-			printf("\toption\ttarget\t'/mnt/%s'\n", basename(pr->dev));
-		}
-	}
-	if (pr->uuid)
-		printf("\toption\tuuid\t'%s'\n", pr->uuid);
-	else
-		printf("\toption\tdevice\t'%s'\n", pr->dev);
-	printf("\toption\tfstype\t'%s'\n", (!strcmp(pr->id->name, "ntfs") ? "ntfs-3g" : pr->id->name));
-	printf("\toption\tenabled\t'0'\n\n");
-
-	return 0;
-}
-
 static int print_block_info(struct probe_info *pr)
 {
 	static char *mp;
@@ -1719,7 +1694,6 @@ static int main_umount(int argc, char **argv)
 
 static int main_detect(int argc, char **argv)
 {
-	struct probe_info *pr;
 
 	cache_load(0);
 	printf("config 'global'\n");
@@ -1729,8 +1703,6 @@ static int main_detect(int argc, char **argv)
 	printf("\toption\tauto_mount\t'1'\n");
 	printf("\toption\tdelay_root\t'5'\n");
 	printf("\toption\tcheck_fs\t'0'\n\n");
-	list_for_each_entry(pr, &devices, list)
-		print_block_uci(pr);
 
 	return 0;
 }
